onnxruntime/core/platform/windows/env.cc

```diff
@@ -124,52 +147,171 @@ class WindowsEnv : public Env {
+    int num_cores = 0;
+    for (const auto& group_info : group_info_vec_) {
+      num_cores += group_info.ActiveProcessorCount;
+    }
+    return num_cores;
+  }
+
+  size_t GetDefaultThreadpoolSetting(std::vector<uint64_t>& affinities) const override {
+    ORT_ENFORCE(sizeof(KAFFINITY) <= sizeof(uint64_t), "KAFFINITY is bigger than uint64_t, cannot save it to an uint64_t vector");
+    affinities.clear();
+    KAFFINITY group_id = 0;
+    for (const auto& group_info : group_info_vec_) {
+      size_t bit_offset = 0;
+      constexpr size_t bit_limit = sizeof(KAFFINITY) << 3;
+      KAFFINITY active_processor_mask = group_info.ActiveProcessorMask;
+      for (int i = 0; i < (group_info.ActiveProcessorCount >> 1); ++i) {
+        KAFFINITY processor_mask = 0;
+        size_t processor_count = 0;
+        while (processor_count < 2 && bit_offset < bit_limit) {
+          KAFFINITY current_bit = BitOne << bit_offset;
+          if (active_processor_mask & current_bit) {
+            processor_mask |= current_bit;
+            processor_count++;
+          }
+          bit_offset++;
+        }
+        affinities.push_back(static_cast<uint64_t>(group_id));
+        affinities.push_back(static_cast<uint64_t>(processor_mask));
+      }
+      group_id++;
+    }
+    if (affinities.empty()) {
+      return std::thread::hardware_concurrency() >> 1;
+    }
+    return affinities.size() >> 1;
+  }
+
+  // The function will go over group_info_vec_ searching for processors between [processor_from, processor_to]
+  // whenever found a match in a group, two things will happen sequentially:
+  // 1. Fill the pair of <group_id, processor_mask> for all matched processor in that group;
+  // 2. Break from the loop to stop searching the next group, this is because windows API will fail if the interval
+  //    spans across group boundaries. 
+  std::pair<KAFFINITY, KAFFINITY> GetGroupAffinity(int processor_from, int processor_to) const {
+    if (processor_from > processor_to) {
+      LOGS_DEFAULT(ERROR) << "Processor <from> must be smaller or equal to <to>";
+      return {0, 0};
+    }
+    int processor_id = 1;
+    int processor_count = 0;
+    uint64_t group_id = 0;
+    uint64_t processor_mask = 0;
+    for (const auto& group_info : group_info_vec_) {
+      for (int i = 0; i < group_info.ActiveProcessorCount; ++i, ++processor_id) {
+        if (processor_id >= processor_from && processor_id <= processor_to) {
+          processor_mask |= BitOne << i;
+          processor_count += 1;
+        }
       }
-      // This is the number of logical processors in the current group
-      return sysInfo.dwNumberOfProcessors;
-    }
-    int processorCoreCount = 0;
-    int count = (int)(returnLength / sizeof(SYSTEM_LOGICAL_PROCESSOR_INFORMATION));
-    for (int i = 0; i != count; ++i) {
-      if (buffer[i].Relationship == RelationProcessorCore) {
-        ++processorCoreCount;
+      // when mask is set, do not visit next group
```

Comment:
> // when mask is set, do not visit next group [](http://example.com/codeflow?start=5&length=45)

This means that we the range returned can be less then asked for? Would it be easier to say I want a processor mask for processors in group X?
In fact, it would probably be always all ones except when limited by processor_to.

Reply:
"This means that we the range returned can be less then asked for? "

- the answer is yes, and when that happens, we will log an error complain about this configuration.

"In fact, it would probably be always all ones except when limited by processor_to"

- yes it is.